protocol/contracts/nexus/Relic.sol

```diff
@@ -0,0 +1,344 @@
+        bytes memory
+    ) public virtual override returns (bytes4) {
+        return this.onERC1155BatchReceived.selector;
+    }
+
+    function onERC721Received(
+        address,
+        address,
+        uint256,
+        bytes memory
+    ) public virtual returns (bytes4) {
+        return this.onERC721Received.selector;
+    }
+
+    function getRelicXP(uint256 _relicId) public view returns (uint256) {
+        require(_exists(_relicId), "This Relic doesn't exist");
+        return relicXP[_relicId];
+    }
+
+    function tokenURI(uint256 _relicId)
+        public
+        view
+        virtual
+        override(ERC721, ERC721URIStorage)
+        returns (string memory)
+    {
+        require(_exists(_relicId));
+        return
+            string(
+                abi.encodePacked(
+                    BASE_URIS[_getRarity(_relicId)],
+                    uint256(enclaves[_relicId]).toString()
+                )
+            );
+    }
+    // get Shard balance for a specific shard Id from a Relic
+    function getBalance(uint256 _relicId, uint256 _tokenId)
+        external
+        view
+        returns (uint256)
+    {
+        require(_exists(_relicId), "This Relic doesn't exist.");
+        return balances[_relicId][_tokenId];
+    }
+
+    function getBalanceBatch(uint256 _relicId, uint256[] memory _tokenId)
+        external
+        view
+        returns (uint256[] memory)
+    {
+        require(_exists(_relicId), "This Relic doesn't exist.");
+
+        uint256[] memory balancesToReturn = new uint256[](_tokenId.length);
+        for (uint256 i = 0; i < _tokenId.length; i++) {
+            balancesToReturn[i] = balances[_relicId][_tokenId[i]];
+        }
+
+        return balancesToReturn;
+    }
+
+    // fetch Rarity and Enclave from a single Relic
+    function getRelicInfos(uint256 _relicId)
+        external
+        view
+        returns (Rarity, Enclave)
+    {
+        return (_getRarity(_relicId), enclaves[_relicId]);
+    }
+
+    //------- Internal -------//
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 firstTokenId,
+        uint256 batchSize
+    ) internal override(ERC721, ERC721Enumerable) whenNotPaused {
+        super._beforeTokenTransfer(from, to, firstTokenId, batchSize);
+    }
+
+    function _burn(uint256 tokenId)
+        internal
+        override(ERC721, ERC721URIStorage)
+    {
+        super._burn(tokenId);
+    }
+
+    function _getRarity(uint256 _relicId)
+        internal
+        view
+        returns (Rarity currRarity)
+    {
+        for (uint256 i = 0; i < 4; i++) {
+            if (relicXP[_relicId] < thresholds[i]) {
+                return Rarity(i);
+            }
+        }
+        return Rarity(4);
+    }
+
+    //------- Owner -------//
+
+    function pause() public onlyOwner {
+        _pause();
+    }
+
+    function unpause() public onlyOwner {
+        _unpause();
+    }
+
+    function safeMint(address to) public onlyOwner {
```

Comment:
This seems unnecessary when you have `mintFromContract()`, and it doesn't set the enclave either?